crates/signs/src/line.rs

```diff
@@ -0,0 +1,268 @@
+const LINE_WIDTH: f32 = 1.;
+/// Offset above mean sea level of the line, stopping z-fighting with the floor.
+const LINE_OFFSET: Vec3 = Vec3::new(0., 1e-3, 0.);
+/// Material configuration used for the lines to the factory spawn point
+const LINE_MATERIAL: LineMaterial = LineMaterial {
+    color: Color::rgba(0.0, 0.5, 0.0, 0.8),
+    pointiness: 2.,
+    speed: 3.,
+    length: 1.,
+    spacing: 0.5,
+    fade: 3.,
+    alpha_mode: AlphaMode::Blend,
+};
+
+pub(crate) struct LinePlugin;
+
+impl Plugin for LinePlugin {
+    fn build(&self, app: &mut App) {
+        app.add_plugin(MaterialPlugin::<LineMaterial>::default())
+            .add_event::<UpdateLineLocationEvent>()
+            .add_event::<UpdateLineEndEvent>()
+            .add_event::<UpdateLineVisibilityEvent>()
+            .add_event::<SpawnLineEvent>()
+            .add_event::<DespawnLineEvent>()
+            .add_system(setup.in_schedule(OnEnter(AppState::InGame)))
+            .add_system(cleanup.in_schedule(OnExit(AppState::InGame)))
+            .add_system(
+                update_line_end
+                    .in_base_set(GameSet::PostUpdate)
+                    .run_if(in_state(AppState::InGame))
+                    .run_if(on_event::<UpdateLineEndEvent>())
+                    .in_set(LinesSet::LineEnd),
+            )
+            .add_system(
+                update_line_location
+                    .in_base_set(GameSet::PostUpdate)
+                    .run_if(in_state(AppState::InGame))
+                    .run_if(on_event::<UpdateLineLocationEvent>())
+                    .in_set(LinesSet::LocationEvents)
+                    .after(LinesSet::LineEnd),
+            )
+            .add_system(
+                update_line_visibility
+                    .in_base_set(GameSet::PostUpdate)
+                    .run_if(in_state(AppState::InGame))
+                    .run_if(on_event::<UpdateLineVisibilityEvent>())
+                    .in_set(LinesSet::VisibilityEvents),
+            )
+            .add_system(
+                spawn_line
+                    .in_base_set(GameSet::PostUpdate)
+                    .run_if(in_state(AppState::InGame))
+                    .run_if(on_event::<SpawnLineEvent>())
+                    .in_set(LinesSet::SpawnLines)
+                    .after(LinesSet::VisibilityEvents),
+            )
+            .add_system(
+                despawn_line
+                    .in_base_set(GameSet::PostUpdate)
+                    .run_if(in_state(AppState::InGame))
+                    .run_if(on_event::<DespawnLineEvent>())
+                    .in_set(LinesSet::SpawnLines)
+                    .after(LinesSet::VisibilityEvents),
+            );
+    }
+}
+
+#[derive(Copy, Clone, Hash, Debug, PartialEq, Eq, SystemSet)]
+enum LinesSet {
+    LineEnd,
+    LocationEvents,
+    VisibilityEvents,
+    SpawnLines,
+}
+
+#[derive(Resource)]
+struct LineMesh(Handle<Mesh>);
+
+fn setup(mut commands: Commands, mut meshes: ResMut<Assets<Mesh>>) {
```

Comment:
Style: I usually put all structs at the top and all systems at the bottom of the module.